draft-ietf-taps-interface.md

```diff
@@ -1780,37 +1772,179 @@ that the other endpoint is done sending on a connection.
 
+Applications can define a Message Framer by creating an object to represent a unique
+definition of a framing protocol. Event handlers for inbound and outbound data
+are registered on this object.
+
+~~~
+framer := NewMessageFramer()
+framer.setEventHandlers()
 ~~~
-Preconnection.DeframeWith(Deframer)
 
-{messageData} := Deframer.Deframe(OctetStream)
+## Adding Message Framers to Connections
+
+The Message Framer object can then be added to one or more Preconnections
+to run on top of transport protocols. Multiple Framers may be added.
+
 ~~~
+Preconnection.AddFramer(framer)
+~~~
+
+When sending Messages, applications can add specific Message
+values to a MessageContext ({{message-props}}) that is intended for a Framer.
+This can be used, for example, to set the type of a Message for a TLV protocol.
+The namespace of values is custom for each unique Message Framer.
+
+~~~
+messageContext := NewMessageContext()
+messageContext.add(framer, key, value)
+Connection.Send(messageData, messageContext)
+~~~
+
+When an application receives a MessageContext in a Receive event,
+it can also look to see if a value was set by a specific Message Framer.
+
+~~~
+messageContext.get(framer, key) -> value
+~~~
+
+## Message Framer Lifetime
+
+When a Connection establishment attempt begins, an event is delivered to
+notify the framer that a new Connection is being created and has a new
+instance of a Message Framer. Similarly, a stop event is delivered when a
+Connection is being torn down.
+
+~~~
+FramerInstance -> Start()
+FramerInstance -> Stop()
+~~~
+
+When Message Framer gets a start message, it sets up its state and then indicates
+to the connection that it is ready to handle sending Messages by calling `ready` on
+the FramerInstance. This call can be delayed in case the protocol needs to do some
+simple handshake exchange before handling application Messages.
+
+~~~
+// instance is a FramerInstance reference
+instance.ready()
+~~~
+
+At any time if the Message Framer encounters a fatal error, it can also cause the Connection
+to fail and provide an error.
+
+~~~
+instance.failed(Error)
+~~~
+
+Before a Message Framer marks itself as ready, it can also dynamically add a protocol or framer
+above it in the stack. This allows protocols like STARTTLS, that need to add TLS conditionally,
+to modify the Protocol Stack based on a handshake result.
+
+~~~
+otherFramer := NewMessageFramer()
+instance.prependProtocol(otherFramer)
+~~~
+
+## Sender-side Message Framing {#send-framing}
+
+Message Framers define an event that gets called on an instance of a
+Framer to handle a new Message being sent by the application.
+
+~~~
+FramerInstance -> NewSentMessage<messageData, messageContext, endOfMessage>
+~~~
+
+Upon receiving this event, a Message Framer implementation is responsible for
+performing any necessary transformations and sending the resulting data to the next
+protocol. Implementations SHOULD ensure that there is a way to pass the original data
```

Comment:
What happens if the final protocol has been reached? How does the message that went through all layers of framers get returned to the connection so it can be send out?

Reply:
The framers are part of the connection protocol stack.. they go down to the transport next.